feat: adding requestor upgrade mode

heyvister1 opened the pull request:
Following PR changes incorporates integration with NVIDIA [maintenance operator](https://github.com/Mellanox/maintenance-operator)

PR include the following changes:
- Restructure upgrade package to accommodate new `requestor` mode. Initially both `inplace`(legacy) as well as `requestor` modes will be supported. The motivation is to use `requestor` and retire upgrade logic from GPU/Network operators.
- Adding upgrade `requestor` package support

almaslennikov commented on hack/crd/bases/maintenance.nvidia.com_nodemaintenances.yaml:
```diff
@@ -0,0 +1,285 @@
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.15.0
+  name: nodemaintenances.maintenance.nvidia.com
```
Why do we need this CRD here?

heyvister1 replied:
It is used by UTs

almaslennikov commented on pkg/upgrade/base/consts.go:
```diff
@@ -39,6 +39,8 @@ const (
 	// Setting this label will trigger setting upgrade state to upgrade-required
 	UpgradeRequestedAnnotationKeyFmt = "nvidia.com/%s-driver-upgrade-requested"
+	// UpgradeRequestorModeAnnotationKeyFmt
+	UpgradeRequestorModeAnnotationKeyFmt = "nvidia.com/%s-driver-upgrade-requestor-mode"
```
Can we add a short README section on the difference between the `legacy` and `requestor` modes?

almaslennikov commented on pkg/upgrade/requestor/upgrade_requestor.go:
```diff
@@ -0,0 +1,247 @@
+	MaintenanceOPFinalizerName      = "maintenance.nvidia.com/finalizer"
+	MaintenanceOPDefaultRequestorID = "nvidia.network.operator"
+	MaintenanceOPEvictionGPU        = "nvidia.com/gpu-*"
+	MaintenanceOPEvictionRDMA       = "nvidia.com/rdma*"
```
Can we add `ResourceName` suffix to the variables' names? It's a little confusing what they are supposed to mean

almaslennikov commented on pkg/upgrade/base/upgrade_common.go:
```diff
@@ -0,0 +1,857 @@
+}
+
+// GetCurrentUnavailableNodes returns all nodes that are not in ready state
+// TODO: Drop ctx as it's not used
```
What's the plan for dropping `ctx`? Can we deprecate this method and add another one without context?
https://go.dev/wiki/Deprecated

almaslennikov commented on pkg/upgrade/upgrade_state.go:
```diff
@@ -79,12 +87,18 @@ func NewClusterUpgradeStateManager(
 	manager := &ClusterUpgradeStateManagerImpl{
 		CommonUpgradeManagerImpl: common,
+		requestor:                request,
 		inplace:                  inplace,
```
do we need to keep both requestor and inplace objects? I think we can have a single `processNodeStateManager` field since they are not used simultaneously and share the same interface

heyvister1 replied:
The motivation is to retire inplace mode, but initially they can be used simultaneously

almaslennikov commented on pkg/upgrade/upgrade_state.go:
```diff
@@ -23,201 +23,76 @@ import (
+// since upgrade pkg is a high level abstaction referencing inplace/requestor (maintenance OP) packages
+type ExtendedUpgradeStateManager interface {
+	// BuildState builds a point-in-time snapshot of the driver upgrade state in the cluster.
+	BuildState(ctx context.Context, namespace string,
```
I think that we should add an extension point at a different level. 

a lot of the files in the base package are only used with the inplace mode (cordon manager, drain manager). Because of this, I think that we should move them to the inplace package and move the `ApplyState` method to the ProcessNodeStateManager interface.

That way, we can have two clearly defined different flows for the inplace and requestor modes.

almaslennikov reviewed (approved):
LGTM. Thank you for addressing the comments!

adrianchiris commented on pkg/upgrade/base/processor/processor.go:
```diff
@@ -0,0 +1,807 @@
+)
+
+// CommonUpgradeStateManager interface is a unified cluster upgrade abstraction for both upgrade modes
+type CommonUpgradeStateManager interface {
```
should we have this interface defined in upgrade_state.go ? since this is a user-facing interface

heyvister1 replied:
Essentially  `CommonUpgradeStateManager ` is embedded under `ClusterUpgradeStateManager`, so user has access a direct to it. The motivation was to put it under base package, since other base providers/managers are using this interface

adrianchiris reviewed (commented):
partial review

adrianchiris commented:
/ok to test

adrianchiris commented on docs/automatic-ofed-upgrade.md:
```diff
@@ -98,4 +98,19 @@ kubectl delete pod -n `kubectl get -A pods --field-selector spec.nodeName=<node_
\ No newline at end of file
+* If after the restart the pod still fails, change the driver version in the CustomResource to the previous or other working version
+
+### New Requestor upgrade mode
```
i would modify this abit.

under Details have a section called: "Upgrade Modes" where you introduce the two modes

then merge the Additional added node upgrade states into "Node upgrade states" with the added information of which states are relevant to which mode WDYT ?

heyvister1 replied:
Good idea. Done

adrianchiris replied:
btw, controllers using this mode would need to watch on NodeMaintenance obj with the provided predicate (requestor/predicate.go) ?

maybe also mention this

heyvister1 replied:
I'll add it

adrianchiris replied:
so we no longer need this section right ?

adrianchiris commented on pkg/upgrade/requestor/client.go:
```diff
@@ -0,0 +1,137 @@
+	err := m.K8sClient.Get(ctx, types.NamespacedName{
+		Name: nodeName, Namespace: m.opts.MaintenanceOPRequestorNS},
+		nm, &client.GetOptions{})
+	if err != nil && !k8serrors.IsNotFound(err) {
```
what if object is not found ?

heyvister1 replied:
Basically node state will stay at `node-maintenance-required`, but I think I should set it again to `upgrade-required` to create new nodeMaintenance obj, and start upgrade again. WDYT?

adrianchiris replied:
please note this is a different discussion where we try to convert to nm at L96 of an "empty" object since we didnt find it (L93)

heyvister1 replied:
Done

adrianchiris commented on pkg/upgrade/requestor/client.go:
```diff
@@ -0,0 +1,137 @@
+	err = m.K8sClient.Create(ctx, nm, &client.CreateOptions{})
+	if err != nil {
+		if k8serrors.IsAlreadyExists(err) {
+			m.Log.V(consts.LogLevelError).Error(err, "nodeMaintenance")
```
is it really an error ?

heyvister1 replied:
I'll change it to Info

adrianchiris commented on pkg/upgrade/requestor/upgrade_requestor.go:
```diff
@@ -0,0 +1,245 @@
+				m.Log.V(consts.LogLevelWarning).Info("missing node annotation", "node", nodeState.Node.Name,
+					"annotations", nodeState.Node.Annotations)
+			}
+			continue
```
if node maintenance is required and node maintenance object is nil, you skip processing it,
should the flow:

1. move to upgrade failed state ?
2. move back to upgrade required ?

currently that node will remain in this state forever ?

heyvister1 replied:
As agreed, node state will move back to `upgrade-required` to initiate new nodeMaintenance obj

heyvister1 replied:
Done

heyvister1 commented on pkg/upgrade/base/drainercordoner/upgrade_suit_test.go:
```diff
@@ -0,0 +1,386 @@
+var cordonManager mocks.CordonManager
+var validationManager mocks.ValidationManager
+var eventRecorder = record.NewFakeRecorder(100)
+
```
done

heyvister1 commented on pkg/upgrade/base/drainercordoner/upgrade_suit_test.go:
```diff
@@ -0,0 +1,386 @@
+	nodeUpgradeStateProvider.
+		On("GetNode", mock.Anything, mock.Anything).
+		Return(
+			func(ctx context.Context, nodeName string) *corev1.Node {
```
done

heyvister1 commented on pkg/upgrade/base/drainercordoner/upgrade_suit_test.go:
```diff
@@ -0,0 +1,386 @@
+	for i := range createdObjects {
+		r := createdObjects[i]
+		key := client.ObjectKeyFromObject(r)
+		err := k8sClient.Get(context.TODO(), key, r)
```
done

heyvister1 commented on pkg/upgrade/base/drainercordoner/upgrade_suit_test.go:
```diff
@@ -0,0 +1,386 @@
+			Annotations: map[string]string{"dummy-key": "dummy-value"},
+		},
+	}
+	Expect(node.Labels).NotTo(BeNil())
```
done

heyvister1 commented on pkg/upgrade/base/processor/processor.go:
```diff
@@ -0,0 +1,807 @@
+}
+
+// GetCurrentUnavailableNodes returns all nodes that are not in ready state
+// TODO: Drop ctx as it's not used
```
done

heyvister1 commented on pkg/upgrade/base/upgrade_common.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+// ClusterUpgradeState contains a snapshot of the driver upgrade state in the cluster
+// It contains driver upgrade policy and mappings between nodes and their upgrade state
```
done

heyvister1 commented on pkg/upgrade/base/util.go:
```diff
@@ -0,0 +1,159 @@
+	return fmt.Sprintf("%sDriverUpgrade", strings.ToUpper(DriverName))
+}
+
+func LogEventf(recorder record.EventRecorder, object runtime.Object, eventType string, reason string, messageFmt string,
```
done

heyvister1 commented on pkg/upgrade/base/processor/processor.go:
```diff
@@ -0,0 +1,807 @@
+
+// ProcessUncordonRequiredNodes processes UpgradeStateUncordonRequired nodes,
+// uncordons them and moves them to UpgradeStateDone state
+func (m *CommonUpgradeManagerImpl) ProcessUncordonRequiredNodes(
```
requestor is also using it to delete nodeMaintenance objects

adrianchiris replied:
indeed but its a different implementation.

adrianchiris replied:
so this needs to move to inplace

adrianchiris commented on pkg/upgrade/requestor/upgrade_requestor.go:
```diff
@@ -0,0 +1,245 @@
+	return nil
+}
+
+// ProcessPostMaintenanceNodes processes UpgradeStatePostMaintenanceRequired
```
you process nodes in NodeMaintenanceRequired state, the name is confusing to me.

why not name this one:  ProcessNodeMaintenanceRequiredNodes() ? similar to other Process<state name> funcs ?

later on if we have a postmaintenance state then this func can move the state machine to post maintenance state instead of pod restart ?

adrianchiris replied:
i think i see what you want to achieve,
the upper level upgrade state manager calls processPostMaintenanceNodes for both inplace and requestor

then accoring to the implementation, different things happen.

maybe have another method that is called `processNodeMaintenanceRequiredNodes()` and call it here ?

heyvister1 replied:
Changed

heyvister1 commented on pkg/upgrade/inplace/upgrade_inplace.go:
```diff
@@ -0,0 +1,119 @@
+
+func (m *UpgradeManagerImpl) ProcessPostMaintenanceNodes(ctx context.Context,
+	currentClusterState *base.ClusterUpgradeState) error {
+	// in future versions we'll remove 'pod-restart-required' and use 'post-maintenance-required' instead...
```
yes. I have added a TODO

adrianchiris commented on pkg/upgrade/requestor/upgrade_requestor.go:
```diff
@@ -0,0 +1,245 @@
+				return fmt.Errorf("failed to update node state. %v", err)
+			}
+			// remove requestor upgrade annotation
+			err := m.NodeUpgradeStateProvider.ChangeNodeUpgradeAnnotation(ctx,
```
why not use SetUpgradeRequestorModeAnnotation to also remove the annot ?

heyvister1 replied:
I will use ChangeNodeUpgradeAnnotation for both cases, I'll delete `SetUpgradeRequestorModeAnnotation`

adrianchiris commented on pkg/upgrade/upgrade_state.go:
```diff
@@ -483,638 +305,44 @@ func (m *ClusterUpgradeStateManagerImpl) ApplyState(ctx context.Context,
-	currentState *ClusterUpgradeState) int {
-	return len(currentState.NodeStates[UpgradeStateFailed])
-}
+	err = m.inplace.ProcessUncordonRequiredNodes(ctx, currentState)
```
so, this is always called IIUC even when requestor mode is enabled.

that means uncordon will be called which is not what we want in case of the use of node maintenance.

adrianchiris replied:
since currentState is not getting re-computed.

i think we need if/else here and handle missing bits in requestor if enabled.

heyvister1 replied:
Thing is that there could be a case where different nodes goes through ProcessUncordonRequiredNodes. Some are going through `inplace` flow, while others (new ones) are going through `requestor` flow. This is why I didn't use if/else.
In addition requestor uncordon node should not go through inplace ProcessUncordonRequiredNodes, since it  is set with a non-matching `uncordon-required` state

adrianchiris replied:
both implementations for PorcessUncordonRequiredNodes process nodes in uncordon-required.

and they are both invoked here in case UseMaintenanceOperator is set.

heyvister1 replied:
Done

heyvister1 commented on pkg/upgrade/base/node_upgrade_state_provider.go:
```diff
@@ -11,7 +11,7 @@ See the License for the specific language governing permissions and
 */
 
-package upgrade
+package base
```
Basically we could have a separate package for `pod_manager`, `safe_driver_load as well `, but lets just keep it as is at the moment

adrianchiris commented on docs/automatic-ofed-upgrade.md:
```diff
@@ -86,6 +89,31 @@ _NOTE: the diagram is outdated_
+New `requestor` upgrade mode is utilizing [NVIDIA maintenance operator](https://github.com/Mellanox/maintenance-operator) nodeMaintenance k8s API objects, to initiate DOCA driver upgrade process.
+Essentially it will retire current upgrade controller (inplace mode) from performing the following node operations: cordon, wait for pods completion, drain, uncordon.
+To enable requestor mode the following environment variable should be enabled `MAINTENANCE_OPERATOR_ENABLED=true`.
+requestor also exposes controller-runtime predicate funtions under `pkg/upgrade/requestor/predicate.go`, embed in your
```
funtions -> functions

... which you should embed ini your controller Watches:
...

heyvister1 replied:
Done

adrianchiris commented on pkg/upgrade/requestor/client.go:
```diff
@@ -0,0 +1,137 @@
+	return nil
+}
+
+// TODO: Check if this is needed
```
do we need this ?

heyvister1 replied:
Removed

adrianchiris commented on pkg/upgrade/requestor/predicate.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+type ConditionChangedPredicate struct {
+	predicate.Funcs
```
for both you should not care imo

heyvister1 replied:
No changes done

adrianchiris commented on pkg/upgrade/requestor/predicate.go:
```diff
@@ -0,0 +1,113 @@
+	condChanged := !reflect.DeepEqual(oldO.Status.Conditions, newO.Status.Conditions)
+	// Check if the object is marked for deletion
+	deleting := len(newO.Finalizers) == 0 && len(oldO.Finalizers) > 0
+	deleting = deleting || !newO.DeletionTimestamp.IsZero()
```
should this be an AND op ?

heyvister1 replied:
Done

heyvister1 commented on pkg/upgrade/requestor/predicate.go:
```diff
@@ -0,0 +1,113 @@
+
+var Scheme = runtime.NewScheme()
+
+// NodeMaintenanceConditionReady designates maintenance operation completed for a designated node
```
Removed

heyvister1 commented on pkg/upgrade/requestor/requestor_suite_test.go:
```diff
@@ -0,0 +1,254 @@
+// These tests use Ginkgo (BDD-style Go testing framework). Refer to
+// http://onsi.github.io/ginkgo/ to learn more about Ginkgo.
+
+var k8sConfig *rest.Config
```
I have removed this unnecessary file

heyvister1 commented on pkg/upgrade/requestor/requestor_suite_test.go:
```diff
@@ -0,0 +1,254 @@
+	return string(b)
+}
+
+func randRange(min, max int) int {
```
I have removed this unnecessary file

heyvister1 commented on pkg/upgrade/requestor/requestor_suite_test.go:
```diff
@@ -0,0 +1,254 @@
+var eventRecorder = record.NewFakeRecorder(100)
+var createdObjects []client.Object
+
+func TestAPIs(t *testing.T) {
```
I have removed requestor_suite_test.go

heyvister1 commented on pkg/upgrade/requestor/client.go:
```diff
@@ -0,0 +1,137 @@
+)
+
+var (
+	DefaultNodeMaintenance maintenancev1alpha1.NodeMaintenance
```
Changed it to be local

heyvister1 commented on pkg/upgrade/requestor/client.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+func (m *UpgradeManagerImpl) NewNodeMaintenance(nodeName string) *maintenancev1alpha1.NodeMaintenance {
+	nm := DefaultNodeMaintenance
```
Done

heyvister1 commented on pkg/upgrade/base/upgrade_common.go:
```diff
@@ -0,0 +1,67 @@
+	Node            *corev1.Node
+	DriverPod       *corev1.Pod
+	DriverDaemonSet *appsv1.DaemonSet
+	NodeMaintenance *unstructured.Unstructured
```
Changes it to be client.Object

adrianchiris commented on pkg/upgrade/base/processor/processor.go:
```diff
@@ -0,0 +1,807 @@
+//	bool: True if the Pod is Orphaned
+//	error: In case of error retrivieng the Revision Hashes
+func (m *CommonUpgradeManagerImpl) podInSyncWithDS(ctx context.Context,
+	nodeState *base.NodeUpgradeState) (bool, bool, error) {
```
@heyvister1 what about this one ?

adrianchiris commented on pkg/upgrade/base/processor/processor.go:
```diff
@@ -0,0 +1,807 @@
+	return m.PodManager.SchedulePodEviction(ctx, &podManagerConfig)
+}
+
+// ProcessPodRestartNodes processes UpgradeStatePodRestartRequirednodes and schedules driver pod restart for them.
```
@heyvister1 what about this one ?

adrianchiris commented on pkg/upgrade/base/processor/processor.go:
```diff
@@ -0,0 +1,807 @@
+		}
+		if !isPodSynced || isOrphaned {
+			// Pods should only be scheduled for restart if they are not terminating or restarting already
+			// To determinate terminating state we need to check for deletion timestamp with will be filled
```
@heyvister1 what about this one ?

adrianchiris reviewed (commented):
minor nits left to address otherwise lgtm

adrianchiris commented:
@heyvister1 please run go mod tidy and commit changes so go-check check passes

heyvister1 commented on pkg/upgrade/requestor/upgrade_requestor.go:
```diff
@@ -0,0 +1,245 @@
+
+//nolint:revive
+type RequestorOptions struct {
+	// UseMaintenanceOperator enables requestor updrade mode
```
Done

heyvister1 commented on pkg/upgrade/requestor/upgrade_requestor.go:
```diff
@@ -0,0 +1,245 @@
+	opts RequestorOptions
+}
+
+// NewClusterUpgradeStateManager creates a new instance of UpgradeManagerImpl
```
Done

heyvister1 commented on pkg/upgrade/requestor/requestor_suite_test.go:
```diff
@@ -0,0 +1,254 @@
+	Expect(err).NotTo(HaveOccurred())
+})
+
+var _ = BeforeEach(func() {
```
I have removed requestor_suite_test.go

heyvister1 commented on pkg/upgrade/requestor/upgrade_requestor.go:
```diff
@@ -0,0 +1,245 @@
+)
+
+const (
+	MaintenanceOPFinalizerName      = "maintenance.nvidia.com/finalizer"
```
Removed `MaintenanceOPFinalizerName`  and added docsting for other exposed consts

heyvister1 commented on pkg/upgrade/base/processor/processor.go:
```diff
@@ -0,0 +1,807 @@
+package processor
```
changed to commonmanager :-)

adrianchiris reviewed (approved):
LGTM, thx for addressing my comments !

adrianchiris commented:
/ok to test

heyvister1 commented:
A kindly reminder for reviewing this PR @tariq1890 @cdesiniotis 
Thanks

tariq1890 commented on pkg/upgrade/base/commonmanager/common_manager.go:
```diff
@@ -0,0 +1,770 @@
+package commonmanager
```
Can we think of a better package name here? `commonmanager/common_manager.go` seems redundant

heyvister1 replied:
You're right, I will think of some other name. Feel free to offer :-)

heyvister1 replied:
`commonmanager` package has been replaced with `upgrade` package

heyvister1 commented on pkg/upgrade/base/consts.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-package upgrade
+package base
```
As you can see, [upgrade](https://github.com/NVIDIA/k8s-operator-libs/pull/86/files#diff-0eb4370a855a7c08a64c3e36b6063996e3607281cbca01f321056bc5299da269L17) package has been preserved, as upper level user facing package.
base pkg was created, since its being referenced by `upgrade`/`inline`/`requestor` packages

tariq1890 replied:
> base pkg was created, to be referenced by upgrade/inline/requestor packages

Can we preserve the overall package structure? I'd prefer if we don't commit any breaking changes to this PR since gpu-operator also depends on the `upgrade` package.  I understand that you want to maintain separate packages for the legacy (in-place) and requestor modes, I am just wondering if it's possible to just use an interface and implementations whilst maintaining existing package structure

heyvister1 replied:
So just to understand, you'd like me to flatten new structure, so all files will be located under upgrade package?

heyvister1 replied:
Done

tariq1890 replied:
Thanks!

cdesiniotis commented on docs/automatic-ofed-upgrade.md:
```diff
@@ -86,6 +89,31 @@ _NOTE: the diagram is outdated_
 ![State change diagram](images/driver-upgrade-state-diagram.png)
 
+#### Upgrade modes
+##### inplace
```
This comment is marked as resolved but I don't see the suggested change being applied.

heyvister1 replied:
This was removed under my last changes. Will fix it

cdesiniotis commented on docs/automatic-ofed-upgrade.md:
```diff
@@ -86,6 +89,31 @@ _NOTE: the diagram is outdated_
 
+#### Upgrade modes
+##### inplace
+inplace (legacy) mode is incorporating full driver upgrade lifecycle, including nodes operations e.g. cordon, pod eviction, drain, uncordon.
```
This comment is marked as resolved but I don't see the suggested change being applied.

cdesiniotis commented on docs/automatic-ofed-upgrade.md:
```diff
@@ -86,6 +89,31 @@ _NOTE: the diagram is outdated_
+to provided `maxParallelUpgrades` under `UpgradePolicy`.
+
+##### requestor
+New `requestor` upgrade mode is utilizing [NVIDIA maintenance operator](https://github.com/Mellanox/maintenance-operator) nodeMaintenance k8s API objects, to initiate DOCA driver upgrade process.
```
This comment is marked as resolved but I don't see the suggested change being applied.

heyvister1 replied:
This was removed under my last changes. Will fix it

cdesiniotis commented on pkg/upgrade/upgrade_state.go:
```diff
@@ -23,191 +23,76 @@ import (
+
+// ClusterUpgradeStateManager is an interface for performing cluster upgrades of driver containers
+type ClusterUpgradeStateManager interface {
+	ExtendedUpgradeStateManager
+	CommonUpgradeStateManager
```
Question -- can you explain at a high-level why these two distinct structs are necessary?

cdesiniotis replied:
If `ExtendedUpgradeStateManager` is not used anywhere but here, would it be better for readability sake to just explicitly define the `BuildState` and `ApplyState` functions here?

heyvister1 replied:
The intention was to decouple `ApplyState/BuildState` from being implemented in `inplace/requestor` modes, and satisfy `ClusterUpgradeStateManager` interface.
Anyway I will remove `ExtendedUpgradeStateManager` and embed its methods under `ClusterUpgradeStateManager`

cdesiniotis commented on pkg/upgrade/upgrade_state.go:
```diff
@@ -278,89 +163,39 @@ func (m *ClusterUpgradeStateManagerImpl) BuildState(ctx context.Context, namespa
+// BuildNodeUpgradeState creates a mapping between a node,
 // the driver POD running on them and the daemon set, controlling this pod
-func (m *ClusterUpgradeStateManagerImpl) buildNodeUpgradeState(
+func (m *ClusterUpgradeStateManagerImpl) BuildNodeUpgradeState(
```
Question: why are we exporting this method now?

cdesiniotis reviewed (commented):
@heyvister1 I made an initial pass. I left some comments / questions but I do not want to block this PR and all the hard work spent on it. From a quick glance, I don't think the changes introduce any breaking changes -- especially for the GPU Operator which will still be using the in-place upgrade mode. However, it would be helpful to verify this by importing these change in the GPU Operator repo, raising a draft PR, and triggering our e2e tests. @heyvister1 would you be willing to do that?

The comments / questions I left on this PR are primarily focused on code structure and readability. I am okay if we identify areas of improvement and refactor in some follow-ups.

heyvister1 commented on pkg/upgrade/upgrade_inplace.go:
```diff
@@ -0,0 +1,145 @@
+	return nil
+}
+
+func (m *InplaceUpgradeManagerImpl) ProcessNodeMaintenanceRequiredNodes(ctx context.Context,
+	currentClusterState *ClusterUpgradeState) error {
+	// TODO: in future versions we'll remove 'pod-restart-required' and use 'post-maintenance-required' instead
+	return m.ProcessPodRestartNodes(ctx, currentClusterState)
+}
```
I will remove `m.ProcessPodRestartNodes` and return it to be under `ApplyState` state machine

heyvister1 commented on pkg/upgrade/upgrade_state.go:
```diff
@@ -23,191 +23,76 @@ import (
-//
-//nolint:interfacebloat
-type ClusterUpgradeStateManager interface {
+// ExtendedUpgradeStateManager interface purpose is to decouple ApplyState implementation from base package
+// since upgrade pkg is a high level abstaction referencing inplace/requestor (maintenance OP) packages
```
I will remove this interface and embed its methods under `ClusterUpgradeStateManager` interface

heyvister1 commented on pkg/upgrade/upgrade_requestor.go:
```diff
@@ -0,0 +1,437 @@
+}
+
+// NewRequestorUpgradeManagerImpl creates a new instance of (requestor) RequestorUpgradeManagerImpl
+func NewRequestorUpgradeManagerImpl(
+	common *CommonUpgradeManagerImpl,
+	opts RequestorOptions) (ProcessNodeStateManager, error) {
```
I will change `RequestorUpgradeManagerImpl` to be `RequestorNodeStateManagerImpl`  and `InplaceUpgradeManagerImpl` accordingly

heyvister1 commented on pkg/upgrade/consts.go:
```diff
@@ -53,6 +55,14 @@ const (
+	// following state designates node operation states (e.g cordon, drain), since they will be handled externally
+	// by maintenance operator
+	UpgradeStateNodeMaintenanceRequired = "node-maintenance-required"
+	// UpgradeStatePostMaintenanceRequired is set when UseMaintenanceOperator is true and maintenance operator exists.
+	// in this case node maintenance operations will be performed externally by maintenance operator.
+	// this state designates requestor to perform post maintenance operations (e.g restart driver's pod etc')
```
Thanks for suggesting, will add this

heyvister1 commented on pkg/upgrade/consts.go:
```diff
@@ -53,6 +55,14 @@ const (
 	// UpgradeStateDrainRequired is set when the node is required to be scheduled for drain. After the drain the state
 	// is changed either to UpgradeStatePodRestartRequired or UpgradeStateFailed
 	UpgradeStateDrainRequired = "drain-required"
+	// UpgradeStateNodeMaintenanceRequired is set when UseMaintenanceOperator is true and maintenance operator exists.
+	// following state designates node operation states (e.g cordon, drain), since they will be handled externally
+	// by maintenance operator
```
Thanks for suggesting, will add this

heyvister1 commented on pkg/upgrade/common_manager.go:
```diff
@@ -0,0 +1,807 @@
+	IsValidationEnabled() bool
+}
+
+// ProcessNodeStateManager interface is used for abstracting both upgrade modes: inplace,
+// requestor (e.g. maintenance OP)
+// Similar node states are used in both modes, while changes are introduced within ApplyState Process<state>
+// methods to support both modes logic
```
Yes, these 3 functions are required for implementing different inplace/requestor logics

heyvister1 commented on pkg/upgrade/common_manager.go:
```diff
@@ -0,0 +1,807 @@
+)
+
+// CommonUpgradeStateManager interface is a unified cluster upgrade abstraction for both upgrade modes
+type CommonUpgradeStateManager interface {
```
UpgradeStateManager is already used in code/comments, I could use `UnifiedUpgradeStateManager`, but at time being lets keep it as is

heyvister1 commented on pkg/upgrade/common_manager.go:
```diff
@@ -0,0 +1,807 @@
+	return ClusterUpgradeState{NodeStates: make(map[string][]*NodeUpgradeState)}
+}
+
+type CommonUpgradeManagerImpl struct {
```
Done

heyvister1 commented:
> @heyvister1 I made an initial pass. I left some comments / questions but I do not want to block this PR and all the hard work spent on it. From a quick glance, I don't think the changes introduce any breaking changes -- especially for the GPU Operator which will still be using the in-place upgrade mode. However, it would be helpful to verify this by importing these changes in the GPU Operator repo, raising a draft PR, and triggering our e2e tests. @heyvister1 would you be willing to do that?
> 
> The comments / questions I left on this PR are primarily focused on code structure and readability. I am okay if we identify areas of improvement and refactor in some follow-ups.

Hey @cdesiniotis Thanks for reviewing this PR. I have addressed your comments and added proper cosmetic changes.
As for testing changes under GPU E2E, I could do that, just let me know if I will be able to trigger GPU OP CI in case I file new PR? Otherwise I'll need to wait for you guys to initiate CI pipeline

heyvister1 commented on .golangci.yaml:
```diff
@@ -104,7 +104,7 @@ linters-settings:
     threshold: 100
   funlen:
     lines: 120
-    statements: 55
```
Had to increase this to accommodate additional states under `ApplyState` method

adrianchiris commented on pkg/upgrade/upgrade_requestor.go:
```diff
@@ -0,0 +1,437 @@
+		return err
+	}
+	// send deletion request assuming maintenance OP will handle actual obj deletion
+	err = m.K8sClient.Delete(ctx, nm)
```
skip delete call if deletion timestamp is set ?

heyvister1 replied:
Done

adrianchiris commented on pkg/upgrade/upgrade_requestor.go:
```diff
@@ -0,0 +1,437 @@
+	err = m.K8sClient.Get(ctx, types.NamespacedName{Name: nodeState.Node.Name,
+		Namespace: m.opts.MaintenanceOPRequestorNS},
+		nm, &client.GetOptions{})
+	if err != nil {
```
maybe ignore not found error ? that way you dont need to check this in the calling function

heyvister1 replied:
Done

adrianchiris commented on pkg/upgrade/upgrade_requestor.go:
```diff
@@ -0,0 +1,437 @@
+		if !ok {
+			return fmt.Errorf("failed to cast object to NodeMaintenance. %v", nodeState.NodeMaintenance)
+		}
+		cond := meta.FindStatusCondition(nm.Status.Conditions, maintenancev1alpha1.ConditionReasonReady)
```
nit: maybe have a helper func: `nodeMaintenanceReady(nm) bool` ?

adrianchiris reviewed (approved):
LGTM for the new layout.

ive left a couple of nits (nothing major)

adrianchiris commented:
/ok to test

heyvister1 commented on pkg/upgrade/upgrade_state.go:
```diff
@@ -23,191 +23,76 @@ import (
-		return nil, fmt.Errorf("error creating k8s client: %v", err)
+		return nil, fmt.Errorf("failed to create commonmanager upgrade state manager. %v", err)
+	}
+	request, err := NewRequestorNodeStateManagerImpl(commonmanager, opts.Requestor)
```
Done

tariq1890 commented:
/ok-to-test 0524ba6caf5a2feb5e6d94a5af060926bcba5f67

tariq1890 commented:
LGTM. Thank you for testing the changes against the gpu-operator repo!

 Let's get the signoff from @cdesiniotis before merging since he has left review comments

heyvister1 commented:
> LGTM. Thank you for testing the changes against the gpu-operator repo!
> 
> Let's get the signoff from @cdesiniotis before merging since he has left review comments

Thanks @tariq1890 and @cdesiniotis
@tariq1890 , can you please approve as well?

adrianchiris commented:
/ok to test

adrianchiris commented:
/ok to test f979aa978cf9184fb4c48914db0c231424c530c2
